tests/functional/extension_template_event_order_test.php

```diff
@@ -0,0 +1,83 @@
+<?php
+/**
+*
+* This file is part of the phpBB Forum Software package.
+*
+* @copyright (c) phpBB Limited <https://www.phpbb.com>
+* @license GNU General Public License, version 2 (GPL-2.0)
+*
+* For full copyright and license information, please see
+* the docs/CREDITS.txt file.
+*
+*/
+
+/**
+* @group functional
+*/
+class phpbb_functional_extension_template_event_order_test extends phpbb_functional_test_case
+{
+	protected $phpbb_extension_manager;
+
+	static private $helper;
+
+	static protected $fixtures = [
+		'./',
+	];
+
+	static public function setUpBeforeClass(): void
+	{
+		parent::setUpBeforeClass();
+
+		self::$helper = new phpbb_test_case_helpers(__CLASS__);
+		self::$helper->copy_ext_fixtures(__DIR__ . '/fixtures/ext/', self::$fixtures);
+	}
+
+	static public function tearDownAfterClass(): void
+	{
+		parent::tearDownAfterClass();
+
+		self::$helper->restore_original_ext_dir();
+	}
+
+	protected function setUp(): void
+	{
+		parent::setUp();
+
+		$this->phpbb_extension_manager = $this->get_extension_manager();
+
+		$this->purge_cache();
+	}
+
+	/**
+	* Check a controller for extension foo/bar.
+	*/
+	public function test_template_event_order()
+	{
+		global $phpbb_root_path;
+
+		$this->phpbb_extension_manager->enable('foo/bar');
+		$this->phpbb_extension_manager->enable('foo/foo');
+		$crawler = self::request('GET', 'index.php');
+		$quick_links_menu = $crawler->filter('ul[role="menu"]')->eq(0);
+		$quick_links_menu_nodes_count = (int) $quick_links_menu->filter('li')->count();
+		// Ensure foo/bar template event goes before foo/foo one
+		$this->assertStringContainsString('FOO_BAR_QUICK_LINK', $quick_links_menu->filter('li')->eq($quick_links_menu_nodes_count - 2)->filter('span')->text());
+		$this->assertStringContainsString('FOO_FOO_QUICK_LINK', $quick_links_menu->filter('li')->eq($quick_links_menu_nodes_count - 1)->filter('span')->text());
+
+		// Change template events order, put foo/bar event after foo/foo one
+		$this->phpbb_extension_manager->disable('foo/bar');
+		$this->assertTrue(copy(__DIR__ . '/fixtures/ext/foo/bar/event/template_event_order_lower.php', $phpbb_root_path . 'ext/foo/bar/event/template_event_order.php'));
+		$this->phpbb_extension_manager->enable('foo/bar');
```

Comment:
I'm unsure how the `template_event_order.php` is initially run, or how `template_event_order_lower.php` is run. Assuming they are, are both tests really being utilized here? Default behavior would by default have `FOO_BAR_QUICK_LINK` before `FOO_FOO_QUICK_LINK`. If anything, the test within `template_event_order_lower.php` should be on foo_foo, since foo_bar is already below foo_foo, making it even lower does nothing and that wouldn't test anything.

Reply:
It works like this:
`template_event_order.php` runs as a regular listener and assertion pass. Then `foo/bar` ext got disabled, `template_event_order_lower.php` got copied into `template_event_order.php`, the ext got enabled again and assertion pass with a different listener priority setting.

> `foo_bar` is below `foo_foo`
Not exactly. Actually we don't know the order as it can be random.

> One of the tests will need to be moved to foo_foo.
Which one?

Reply:
> Not exactly. Actually we don't know the order as it can be random.
The way it works by default is that template events execute in alphabetical order of namespace, so `foo_boo` would execute before `foo_foo`.

> Which one?
The `template_event_order_lower.php` event should be moved to `foo_foo` since by default, `foo_foo` would be executed last anyways.
`template_event_order.php` should stay on `foo_bar` since moving it after `foo_foo` would be changed behavior and would test the extension.

Reply:
Okay, so one of tests seems to be useless at al, does it? If `foo/foo` goes after `foo/bar` by default, only 1 test for moving it in reverted order is needed (no matter a test moving `foo/bar` down or a test moving `foo/foo` up).
Or I didn't get your thought otherwise.

Reply:
(I wrote the following edit: before seeing your latest comment):

If you do a `print_r` on the `get_phpbb_extensions()` call, you'll see that it is in alphabetical order of `$ext_namespace`s. Here's a gist with my output:
https://gist.github.com/toxyy/c6266a05c79a75e71e0ae85aba9f653a

So by default `foo_foo` will always be after `foo_bar`, hence the request to change the test. Currently the test to place `foo_bar` before `foo_foo` will test for nothing.

---

> Okay, so one of tests seems to be useless at al, does it?

Could be that only one test is necessary to see if the event indeed moves a template event as it should. Just depends on if you want to test moving an event both after it would otherwise execute and before it would otherwise execute. If you think that only having one is sufficient that's fine.

Reply:
Fair enough, got it. So 1 test for moving thing up and another for moving it down. Ok.

Reply:
Thanks for all your help with getting this done, I really appreciate it

Reply:
Technically there is a third scenario: setting a `$priority_key` in the middle of `$template_events` to an existing key. I made a comment on my review below that this should already result in expected behavior: the newly added event will compile after the previously added one for the same key. I don't believe another test for this case will be necessary, but I did want to state it.

Reply:
Resolving this since it's on outdated code